File: web/src/lib/airlines.ts
"use client";
// Airline directory (public/airlines.json, compiled from engine data) and
// deep link building: open an airline's own search pre-filled with a route.

import { useEffect, useState } from "react";

export type Airline = {
  iata: string;
  icao?: string;
  name: string;
  country: string;
  regions: string[];
  category: "full_service" | "low_cost" | "ultra_low_cost" | "hybrid" | "regional" | "premium_leisure";
  tags: string[];
  alliance: "star" | "oneworld" | "skyteam" | null;
  hubs: string[];
  website: string;
  search_url: string;
  deeplink: string | null;
  date_format?: string;
  oneway_deeplink?: string;
  deeplink_verified?: boolean;
  logo: string;
  notes?: string;
};

export const REGIONS: { id: string; label: string }[] = [
  { id: "global", label: "Global networks" },
  { id: "nordics", label: "Nordics" },
  { id: "europe", label: "Europe" },
  { id: "us_domestic", label: "US domestic" },
  { id: "north_america", label: "Canada" },
  { id: "mexico", label: "Mexico" },
  { id: "central_america_caribbean", label: "Central America and Caribbean" },
  { id: "south_america", label: "South America" },
  { id: "middle_east", label: "Middle East" },
  { id: "africa", label: "Africa" },
  { id: "asia", label: "Asia" },
  { id: "oceania", label: "Oceania" },
];

export const CATEGORY_LABEL: Record<Airline["category"], string> = {
  full_service: "Full service",
  low_cost: "Low cost",
  ultra_low_cost: "Ultra low cost",
  hybrid: "Hybrid",
  regional: "Regional",
  premium_leisure: "Premium leisure",
};

let cache: Airline[] | null = null;
let pending: Promise<Airline[]> | null = null;

export function loadAirlines(): Promise<Airline[]> {
  if (cache) return Promise.resolve(cache);
  pending ??= fetch("/airlines.json")
    .then((r) => r.json())
    .then((d: { airlines: Airline[] }) => (cache = d.airlines));
  return pending;
}

export function useAirlines() {
  const [list, setList] = useState<Airline[] | null>(cache);
  useEffect(() => {
    if (!cache) loadAirlines().then(setList).catch(() => setList([]));
  }, []);
  return list;
}

export function airlineByCode(list: Airline[] | null, code: string) {
  return list?.find((a) => a.iata === code) ?? null;
}

const MON = ["JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOV", "DEC"];
const MON_SHORT = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];

function fmt(iso: string, f = "YYYY-MM-DD") {
  const [y, m, d] = iso.slice(0, 10).split("-");
  switch (f) {
    case "YYYYMMDD":
      return `${y}${m}${d}`;
    case "MM/DD/YYYY":
      return `${m}/${d}/${y}`;
    case "DD/MM/YYYY":
      return `${d}/${m}/${y}`;
    case "DD.MM.YYYY":
      return `${d}.${m}.${y}`;
    case "MMM DD, YYYY":
      return encodeURIComponent(`${MON_SHORT[Number(m) - 1]} ${d}, ${y}`);
    default:
      return `${y}-${m}-${d}`;
  }
}

export type RouteQuery = { origin: string; destination: string; depart: string; ret?: string | null; adults?: number };

// The airline's own search, pre-filled when we know its link format, else its
// plain search page. Returns [url, prefilled].
export function airlineLink(a: Airline, q?: RouteQuery | null): [string, boolean] {
  if (!q || !a.deeplink) return [a.search_url || a.website, false];
  const tpl = !q.ret && a.oneway_deeplink ? a.oneway_deeplink : a.deeplink;
  const f = a.date_format;
  const vals: Record<string, string> = {
    origin: q.origin,
    destination: q.destination,
    adults: String(q.adults ?? 1),
    depart: fmt(q.depart, f),
    return: q.ret ? fmt(q.ret, f) : "",
    depart_mon: MON[Number(q.depart.slice(5, 7)) - 1],
    depart_day: q.depart.slice(8, 10),
    return_mon: q.ret ? MON[Number(q.ret.slice(5, 7)) - 1] : "",
    return_day: q.ret ? q.ret.slice(8, 10) : "",
  };
  return [tpl.replace(/\{(\w+)\}/g, (_, k: string) => vals[k] ?? ""), true];
}

File: engine/src/flightscout/airlines.py
"""Airline directory (data/airlines.json) and deep links into each airline's
own search, pre-filled with a route when the airline supports it. Mirrors
web/src/lib/airlines.ts."""

from __future__ import annotations

import json
from datetime import date
from functools import lru_cache
from importlib import resources
from urllib.parse import quote

REGIONS = {
    "global": "Global networks", "nordics": "Nordics", "europe": "Europe", "us_domestic": "US domestic",
    "north_america": "Canada", "mexico": "Mexico", "central_america_caribbean": "Central America and Caribbean",
    "south_america": "South America", "middle_east": "Middle East", "africa": "Africa", "asia": "Asia",
    "oceania": "Oceania",
}
CATEGORIES = {
    "full_service": "Full service", "low_cost": "Low cost", "ultra_low_cost": "Ultra low cost",
    "hybrid": "Hybrid", "regional": "Regional", "premium_leisure": "Premium leisure",
}
_MON = ["JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOV", "DEC"]
_MON_SHORT = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]


@lru_cache
def all_airlines() -> list[dict]:
    raw = resources.files("flightscout.data").joinpath("airlines.json").read_text("utf-8")
    return json.loads(raw)["airlines"]


def get(code: str) -> dict | None:
    code = code.upper()
    return next((a for a in all_airlines() if a["iata"] == code), None)


def find(query: str = "", region: str | None = None, category: str | None = None,
         alliance: str | None = None) -> list[dict]:
    q = query.strip().lower()
    out = []
    for a in all_airlines():
        if region and region not in a.get("regions", []):
            continue
        if category:
            cats = {"low_cost", "ultra_low_cost", "hybrid"} if category == "budget" else {category}
            if a.get("category") not in cats:
                continue
        if alliance and (a.get("alliance") or "none") != alliance:
            continue
        if q and not (q == a["iata"].lower() or q in a["name"].lower() or any(q in t.lower() for t in a.get("tags", []))):
            continue
        out.append(a)
    return out


def _fmt(d: date, f: str | None) -> str:
    if f == "YYYYMMDD":
        return d.strftime("%Y%m%d")
    if f == "MM/DD/YYYY":
        return d.strftime("%m/%d/%Y")
    if f == "DD/MM/YYYY":
        return d.strftime("%d/%m/%Y")
    if f == "DD.MM.YYYY":
        return d.strftime("%d.%m.%Y")
    if f == "MMM DD, YYYY":
        return quote(f"{_MON_SHORT[d.month - 1]} {d.day:02d}, {d.year}")
    return d.isoformat()


def link(a: dict, origin: str | None = None, destination: str | None = None, depart: date | None = None,
         ret: date | None = None, adults: int = 1) -> tuple[str, bool]:
    """(url, prefilled). Falls back to the airline's search page."""
    if not (origin and destination and depart) or not a.get("deeplink"):
        return a.get("search_url") or a.get("website"), False
    tpl = a["oneway_deeplink"] if (not ret and a.get("oneway_deeplink")) else a["deeplink"]
    f = a.get("date_format")
    vals = {
        "origin": origin.upper(), "destination": destination.upper(), "adults": str(adults),
        "depart": _fmt(depart, f), "return": _fmt(ret, f) if ret else "",
        "depart_mon": _MON[depart.month - 1], "depart_day": f"{depart.day:02d}",
        "return_mon": _MON[ret.month - 1] if ret else "", "return_day": f"{ret.day:02d}" if ret else "",
    }
    for k, v in vals.items():
        tpl = tpl.replace("{" + k + "}", v)
    return tpl, True
